File: src/org/simbrain/network/neuron_update_rules/SpikingThresholdRule.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.neuron_update_rules;

import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.SpikingNeuronUpdateRule;
import org.simbrain.network.neuron_update_rules.interfaces.NoisyUpdateRule;
import org.simbrain.util.UserParameter;
import org.simbrain.util.math.ProbDistributions.UniformDistribution;
import org.simbrain.util.math.ProbabilityDistribution;

import java.util.Random;

/**
 * A simple spiking neuron that fires when weighted inputs exceed a threshold.
 * TODO: Has no documentation.
 */
public class SpikingThresholdRule extends SpikingNeuronUpdateRule implements NoisyUpdateRule {

    /**
     * Threshold.
     */
    @UserParameter(
            label = "Threshold",
            description = "Input value above which the neuron spikes.",
            defaultValue = "0.5", order = 1)
    private double threshold = .5;

    /**
     * The noise generating randomizer.
     */
    private ProbabilityDistribution noiseGenerator = UniformDistribution.create();

    /**
     * Whether or not to add noise to the inputs .
     */
    private boolean addNoise;

    @Override
    public SpikingThresholdRule deepCopy() {
        SpikingThresholdRule neuron = new SpikingThresholdRule();
        neuron.setThreshold(getThreshold());
        return neuron;
    }

    @Override
    public void update(Neuron neuron) {
        final double input = neuron.getInput() + (addNoise ? noiseGenerator.getRandom() : 0);
        if (input >= threshold) {
            neuron.setSpkBuffer(true);
            setHasSpiked(true, neuron);
            neuron.setBuffer(1);
        } else {
            neuron.setSpkBuffer(false);
            setHasSpiked(false, neuron);
            neuron.setBuffer(0); // Make this a separate variable?
        }

    }

    /**
     * {@inheritDoc}
     */
    @Override
    public double getRandomValue() {
        Random rand = new Random();
        return rand.nextBoolean() ? 1 : 0;
    }

    /**
     * @return the threshold
     */
    public double getThreshold() {
        return threshold;
    }

    /**
     * @param threshold the threshold to set
     */
    public void setThreshold(double threshold) {
        this.threshold = threshold;
    }

    @Override
    public String getName() {
        return "Spiking Threshold";
    }

    @Override
    public ProbabilityDistribution getNoiseGenerator() {
        return noiseGenerator;
    }

    @Override
    public void setNoiseGenerator(final ProbabilityDistribution noise) {
        this.noiseGenerator = noise;
    }

    @Override
    public boolean getAddNoise() {
        return addNoise;
    }

    @Override
    public void setAddNoise(boolean noise) {
        this.addNoise = noise;
    }
}

File: src/main/kotlin/org/simbrain/network/updaterules/SpikingThresholdRule.kt
package org.simbrain.network.updaterules

import org.simbrain.network.core.Layer
import org.simbrain.network.core.Network
import org.simbrain.network.core.Neuron
import org.simbrain.network.core.SpikingNeuronUpdateRule
import org.simbrain.network.updaterules.interfaces.NoisyUpdateRule
import org.simbrain.network.util.SpikingMatrixData
import org.simbrain.network.util.SpikingScalarData
import org.simbrain.util.UserParameter
import org.simbrain.util.stats.ProbabilityDistribution
import org.simbrain.util.stats.distributions.UniformRealDistribution
import java.util.*

/**
 * A simple spiking neuron that fires when weighted inputs exceed a threshold.
 * When spiking activation is 1, else it is 0.
 * TODO: Has no documentation.
 */
open class SpikingThresholdRule : SpikingNeuronUpdateRule<SpikingScalarData, SpikingMatrixData>(), NoisyUpdateRule {
    /**
     * Threshold.
     */
    @JvmField
    @UserParameter(
        label = "Threshold",
        description = "Input value above which the neuron spikes.",
        increment = .1,
        order = 1
    )
    var threshold: Double = .5

    /**
     * The noise generating randomizer.
     */
    override var noiseGenerator: ProbabilityDistribution = UniformRealDistribution()

    /**
     * Whether or not to add noise to the inputs .
     */
    override var addNoise: Boolean = false

    override fun deepCopy(): SpikingThresholdRule {
        val neuron = SpikingThresholdRule()
        neuron.threshold = threshold
        return neuron
    }

    context(Network)
    override fun apply(layer: Layer, dataHolder: SpikingMatrixData) {
        for (i in 0 until layer.outputSize()) {
            if (spikingThresholdRule(layer.inputs[i, 0])) {
                dataHolder.setHasSpiked(i, true, time)
                layer.outputs[i, 0] = 1.0
            } else {
                dataHolder.setHasSpiked(i, false, time)
                layer.outputs[i, 0] = 0.0
            }
        }
    }

    context(Network)
    override fun apply(neuron: Neuron, data: SpikingScalarData) {
        if (spikingThresholdRule(neuron.input)) {
            neuron.isSpike = true
            neuron.activation = 1.0
        } else {
            neuron.isSpike = false
            neuron.activation = 0.0 // Make this a separate variable?
        }
    }

    fun spikingThresholdRule(`in`: Double): Boolean {
        val input = `in` + (if (addNoise) noiseGenerator.sampleDouble() else 0.0)
        return if (input >= threshold) {
            true
        } else {
            false
        }
    }

    override val randomValue: Double
        get() {
            val rand = Random()
            return if (rand.nextBoolean()) 1.0 else 0.0
        }

    override val name: String
        get() = "Spiking Threshold"
}